src/robin_dict.jl

```diff
@@ -0,0 +1,593 @@
+    @inbounds h.keys[index_curr] = ckey
+    @inbounds h.hashes[index_curr] = chash
+    
+    @assert probe_current >= 0
+    
+    h.maxprobe = max(h.maxprobe, probe_current)
+    if h.idxfloor == 0
+        h.idxfloor = index_curr
+    else
+        h.idxfloor = min(h.idxfloor, index_curr)
+    end
+    return index_curr
+end
+
+#rehash! algorithm
+function rehash!(h::RobinDict{K,V}, newsz = length(h.keys)) where {K, V}
+    oldk = h.keys
+    oldv = h.vals
+    oldh = h.hashes
+    sz = length(oldk)
+    newsz = _tablesz(newsz)
+    h.totalcost += 1
+    if h.count == 0
+        resize!(h.keys, sz)
+        resize!(h.vals, sz)
+        resize!(h.hashes, newsz)
+        fill!(h.hashes, 0)
+        h.count = 0
+        h.maxprobe = 0
+        h.totalcost = 0
+        h.idxfloor = 0
+        return h
+    end
+
+    h.keys = Vector{K}(undef, newsz)
+    h.vals = Vector{V}(undef, newsz)
+    h.hashes = zeros(UInt32,newsz)
+    totalcost0 = h.totalcost
+    h.count = 0
+    h.maxprobe = 0
+    h.idxfloor = 0
+
+    for i = 1:sz
+        @inbounds if oldh[i] != 0
+            k = oldk[i]
+            v = oldv[i]
+            rh_insert!(h, k, v)
+        end
+    end
+    h.totalcost = totalcost0
+    return h
+end
+
+function sizehint!(d::RobinDict, newsz)
+    newsz = _tablesz(newsz*2)  # *2 for keys and values in same array
+    oldsz = length(d.keys)
+    # grow at least 25%
+    if newsz < (oldsz*5)>>2
+        return d
+    end
+    rehash!(d, newsz)
+end
+
+@propagate_inbounds isslotfilled(h::RobinDict, index) = (h.hashes[index] != 0)
+@propagate_inbounds isslotempty(h::RobinDict, index) = (h.hashes[index] == 0)
+
+
+function setindex!(h::RobinDict{K,V}, v0, key0) where {K, V}
+    key = convert(K, key0)
+    isequal(key, key0) || throw(ArgumentError("$key0 is not a valid key for type $K"))
+    _setindex!(h, key, v0)
+end
+
+function _setindex!(h::RobinDict{K,V}, key::K, v0) where {K, V}
+    v = convert(V, v0)
+    sz = length(h.keys)
+    (h.count > ROBIN_DICT_LOAD_FACTOR * sz) && rehash!(h, sz<<2)
+    index = rh_insert!(h, key, v)
+    @assert index > 0
+    h.totalcost += 1
+    h
+end
+
+isempty(d::RobinDict) = (d.count == 0)
+length(d::RobinDict) = d.count
+
+"""
+    empty!(collection) -> collection
+
+Remove all elements from a `collection`.
+
+# Examples
+```jldoctest
+julia> A = RobinDict("a" => 1, "b" => 2)
+RobinDict{String,Int64} with 2 entries:
+  "b" => 2
+  "a" => 1
+
+julia> empty!(A);
+
+julia> A
+RobinDict{String,Int64} with 0 entries
+```
+"""
+function empty!(h::RobinDict{K,V}) where {K, V}
+    sz = length(h.keys)
+    empty!(h.hashes)
+    empty!(h.keys)
+    empty!(h.vals)
+    resize!(h.keys, sz)
+    resize!(h.vals, sz)
+    resize!(h.hashes, sz)
+    h.count = 0
+    h.maxprobe = 0
+    h.totalcost = 0
+    h.idxfloor = 0
+    return h
+end
+ 
+function rh_search(h::RobinDict{K, V}, key::K) where {K, V}
+    sz = length(h.keys)
+    chash = hash_key(key)
+    index = desired_index(chash, sz)
+    cdibs = 0
```

Comment:
This would be the place for prefetching `h.keys[index]` and `h.vals[index]`, via
`prefetch_r(pointer(h.keys, index))`, using e.g.:
```
@inline function _prefetchr(p::Ptr)
    p = reinterpret(UInt, p)
    if UInt === UInt64
        Core.Intrinsics.llvmcall(("declare void @llvm.prefetch(i8*, i32, i32, i32)", """
    %ptr = inttoptr i64 %0 to i8*
    call void @llvm.prefetch(i8* %ptr, i32 0, i32 3, i32 1)
    ret void
    """), Nothing, Tuple{UInt}, p)
    else
        Core.Intrinsics.llvmcall(("declare void @llvm.prefetch(i8*, i32, i32, i32)", """
    %ptr = inttoptr i32 %0 to i8*
    call void @llvm.prefetch(i8* %ptr, i32 0, i32 3, i32 1)
    ret void
    """), Nothing, Tuple{UInt}, p)
    end
end
```
For a dict with low load factor, this is unneeded: The branch predictor will speculatively execute the loop and prefetch the relevant addresses, spectre-style, while waiting for the `h.hashes[index]` cache miss to resolve.

I am unsure about high loads: When the branch predictor guesses that we won't have a match, then the transient execution might run in a look that fails to fire off the memory load.

Reply:
This seems likely far too late for prefetch to make a difference. You would have needed to call it around a hundred cycles ago for it to be expected to reduce latency. Putting it here, mere cycles away from the expected load, seems likely to just slow down execution (it'll waste one of the a load ports and an instruction micro-op). https://stackoverflow.com/a/3123012

We might be able to move the actual load here, but currently the intent is to follow the implementation in base reasonably closely, for more fair benchmarking.

For this PR, I'd like to see it get merged fairly quickly (after a thorough review for correctness). Then we can work on additional performance evaluation and other trade-offs, relative to the initial version as a reference point.

Reply:
+1 to getting the PR merged quickly and leaving performance optimizations for later, and only merging them if we have an example where they make a difference.

Re "too late for prefetch": We want to fire off the memory load as early as possible. This means firing it off as soon as we have computed the hash, i.e. simultaneously with the memory load for `h.hashes[index]`, instead of waiting until `h.hashes[index]` has arrived and we have compared it to the target `chash`. As I said, this should not matter for low load factors, because speculative execution should fire off the same memory load anyway while waiting for the hash comparison: Spectre-style prefetch. However, if the branch predictor guesses that none of the hashes match, then we risk waiting for a second memory round-trip. Or am I misunderstanding something here?